rfcs/text/0002_encrypted_attributes.md

```diff
@@ -0,0 +1,228 @@
+);
+
+// savedObject = { 
+//   id: 'dd9750b9-ef0a-444c-8405-4dfcc2e9d670',
+//   type: 'server-action',
+//   name: 'my-server-action',
+//   credentials: { username: 'some-user', password: 'some-password' },
+//   apiKey: 'dGhpcyBpcyBub3QgYSByZWFsIHRva2VuIGJ1dCBpdCBpcyBvb',
+// };
+```
+
+# Motivation
+
+Main motivation is the storage and usage of third-party credentials for use with
+the action service to do notifications. Also perform other types integrations,
+call webhooks using tokens.
+
+# Detailed design
+
+In order for this to be in basic it needs to be done as a wrapper around the
+saved object client. This can be added from the `x-pack` plugin.
+
+## General
+
+To be able to manage saved objects with encrypted attributes from any plugin one should
+do the following:
+
+1. Define `encrypted_saved_objects` plugin as a dependency.
+2. Add attributes to be encrypted in `mappings.json` file for the respective saved object type. These attributes should
+always have a `binary` type since they'll contain encrypted content as a `Base64` encoded string and should never be
+searchable or analyzed. This makes defining of attributes that require encryption explicit and auditable, and significantly
+simplifies implementation:
+```json
+{
+ "server-action": {
+   "properties": {
+     "name": { "type": "keyword" },
+     "credentials": { "type": "binary" },
+     "apiKey": { "type": "binary" }
+   }
+ }
+}
+```
+3. Register saved object type and attributes that should be encrypted with `encrypted_saved_objects` plugin:
+```typescript
+server.plugins.encrypted_saved_objects.registerType({
+  type: 'server-action',
+  attributesToEncrypt: new Set(['credentials', 'apiKey']),
+  // Optional list of attributes that might not contain user specified data.
+  attributesToExcludeFromAAD: new Set(['temporaryState']),
+});
+```
+
+Since `encrypted_saved_objects` adds its own wrapper (`EncryptedSavedObjectsClientWrapper`) into `SavedObjectsClient`
+wrapper chain consumers will be able to create, update, delete and retrieve saved objects using standard Saved Objects API.
+Two main responsibilities of the wrapper are:
+
+* It encrypts attributes that are supposed to be encrypted during `create`, `bulkCreate` and `update` operations
+* It strips encrypted attributes from **any** saved object returned from the Saved Objects API
+
+As noted above the wrapper is stripping encrypted attributes from saved objects returned from the API methods, that means
+that there is no way at all to retrieve encrypted attributes using standard Saved Objects API unless `encrypted_saved_objects`
+plugin is disabled. This potentially can lead to the situation when consumer retrieves saved object, updates its non-encrypted
+properties and passes that same object to the `update` Saved Objects API method without re-defining encrypted attributes.
+At this stage we consider this as a developer mistake and don't prevent it from happening in any way apart from logging
```

Comment:
What is the actual consequence of this mistake? Do the encrypted fields get overwritten to nothing or just ignored and untouched? If the latter, I don't even think we need to warn via logging. If the former, than I think we should be throwing an error rather than logging.

Reply:
> Do the encrypted fields get overwritten to nothing or just ignored and untouched?

We use `_update` API under the hood right now, so we'll have a partial update and existing encrypted fields will stay untouched. __But__ if attributes that are being updated are included into AAD then after update one wouldn't be able to decrypt content anymore.

That's another reason why we have `attributesToExcludeFromAAD`, if you want to allow user (that isn't allowed to decrypt attributes) to modify some of the non-encrypted properties of the saved object with encrypted attributes (e.g. change title, but not credentials) then you'd exclude these attributes from AAD so that updates don't matter to decryption/encryption.

So we have two `update` cases here:

* without attributes to be encrypted **and with** with some of the attributes that contribute to AAD
* without attributes to be encrypted **and without** any of the attributes that contribute to AAD (e.g only with attributes mentioned in `attributesToExcludeFromAAD`)

In the latter case we may or may not "debug" log, I think it's useful, but don't have a strong opinion and can remove `debug` log if you think it's not needed.

But the former case is tricky and can have several outcomes:
* all or some attributes that contribute to AAD are included **and** they have the same content ---> decryption will still work, it can potentially be a valid use case
* all or some attributes that contribute to AAD are included **but** some of them have a different content ---> decryption won't work anymore

The problem is that plugin can't know whether or not content of the attributes has changed. So throwing here may be too restrictive, depending on what we think are valid use cases.

Am I missing or overcomplicating something here @kobelb @epixa ?

Reply:
> Am I missing or overcomplicating something here @kobelb @epixa ?

You've summarized the situation quite well! Like you've illustrated, it really becomes quite complicated and I think the approach which you've outlined in your RFC is a good middle-ground for the initial implementation.

Reply:
Sounds good